crates/handlers/src/rate_limit.rs

```diff
@@ -0,0 +1,183 @@
+// Copyright 2024 The Matrix.org Foundation C.I.C.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::{net::IpAddr, sync::Arc};
+
+use governor::{clock::QuantaClock, state::keyed::DashMapStateStore, Quota, RateLimiter};
+use mas_data_model::User;
+use nonzero_ext::nonzero;
+use ulid::Ulid;
+
+const PASSWORD_CHECK_FOR_REQUESTER_QUOTA: Quota = Quota::per_minute(nonzero!(3u32));
+const PASSWORD_CHECK_FOR_USER_QUOTA: Quota = Quota::per_hour(nonzero!(1800u32));
+
+#[derive(Debug, Clone, Copy, thiserror::Error)]
+pub enum PasswordCheckLimitedError {
+    #[error("Too many password checks for requester {0}")]
+    Requester(RequesterFingerprint),
+
+    #[error("Too many password checks for user {0}")]
+    User(Ulid),
+}
+
+/// Key used to rate limit requests per requester
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
+pub struct RequesterFingerprint {
+    ip: Option<IpAddr>,
+}
+
+impl std::fmt::Display for RequesterFingerprint {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        if let Some(ip) = self.ip {
+            write!(f, "{ip}")
+        } else {
+            write!(f, "(NO CLIENT IP)")
+        }
+    }
+}
+
+impl RequesterFingerprint {
+    /// An anonymous key with no IP address set. This should not be used in
+    /// production, and we should warn users if we can't find their client IPs.
+    pub const EMPTY: Self = Self { ip: None };
+
+    /// Create a new anonymous key with the given IP address
+    #[must_use]
+    pub const fn new(ip: IpAddr) -> Self {
+        Self { ip: Some(ip) }
+    }
+}
+
+/// Rate limiters for the different operations
+#[derive(Debug, Clone, Default)]
+pub struct Limiter {
+    inner: Arc<LimiterInner>,
+}
+
+type KeyedRateLimiter<K> = RateLimiter<K, DashMapStateStore<K>, QuantaClock>;
+
+#[derive(Debug)]
+struct LimiterInner {
+    password_check_for_requester: KeyedRateLimiter<RequesterFingerprint>,
+    password_check_for_user: KeyedRateLimiter<Ulid>,
+}
+
+impl Default for LimiterInner {
+    fn default() -> Self {
+        Self {
+            password_check_for_requester: RateLimiter::keyed(PASSWORD_CHECK_FOR_REQUESTER_QUOTA),
```

Comment:
can't remember for sure but I think you want to call `retain_recent` on these ratelimiters occasionally to perform housekeeping.

Reply:
Good call, I did that in 5b1c9b45